### PR TITLE
Add fixedEntryModificationTime attribute to AbstractZipArchiver 

### LINKED ISSUE #48: Add ability to NOT write timestamps for zip file entries
To be discussed:
Thinking in reproducible builds, It would be great if we could see build tools that use plexus-archiver as pure transformers/functions - given the same input - will always produce the same output (at least if they are used synchronous and always provide the same input in the same sequence). This is however currently not possible using maven-archiver (transitively using plexus-archiver) because the AbstractZipArchiver will always create ZipEntry timestamps. I wish there was a property to `AbstractZipArchiver` like `private boolean createZipEntryTime = true;` to turn the setTime behaviour off.

I'd be happy to submit a PR if this feature is desirable (is there maybe something in the jar spec that insists on these fields? I have not found any reference).

### PR DESCRIPTION
**Pull request by danielwegener**
... to override zipEntry times globally. closes #48

### CODE REVIEW

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java [hunk 1]
```diff
@@ -578,7 +597,12 @@ public InputStream get()
 
     private void setTime( java.util.zip.ZipEntry zipEntry, long lastModified )
     {
-        zipEntry.setTime( lastModified + ( isJava7OrLower ? 1999 : 0 ) );
+
+        if (fixedEntryModificationTime != null) {
+            zipEntry.setTime(Math.max(MIN_MS_DOS_TIME, fixedEntryModificationTime));
```
Comment (danielwegener) [inline src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: Other values like 0L produce weird times when reading (seen something like 8. Nov 1979, which normally cannot be expressed by MS-DOS time). Not sure if we should do the round-up here.

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java [hunk 2]
```diff
@@ -212,6 +223,14 @@ public boolean isFilesonly()
         return doFilesonly;
     }
 
+    public Long getFixedEntryModificationTime() {
+        return fixedEntryModificationTime;
+    }
+
+    public void setFixedEntryModificationTime(Long fixedEntryModificationTime) {
```
Comment (krosenvold) [inline src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: I really think this should be some kind of Date object, not a long. I am also contemplating if this setter should be on AbstractArchiver, others could just throw UnsupportedOepratiionException

Comment (plamentotev) [inline src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: Actually would it be even better if it works the same way as forced file/dir modes? I mean they are logically related - forced file/dir modes are options to override the file permissions. This is option to override the mast modified date.

Comment (danielwegener) [inline src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: Summarizing:

- a method `setModificationTime(Date modificationTime)` on `Archiver` (+getter)
- a field `private Date forcedModificationTime = null` on `AbstractArchiver`
- The field is only used by TarArchiver, AbstractZipArchiver and
- Question then: do we also need overloaded methods for addFile/addResource to allow overriding the modificationTime on an entry basis (and having a `modificationTime` field in `ArchiveEntry`)?

Comment (krosenvold) [inline src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: It seems to me like ArchiveEntry needs modifcationTime with that is overridden much the same way as the "mode" field. As for per-entry modifiers I suggest you only add the ones you need for the patch.

**Comment by michael-o** [comment]
Pre-Java 7 is gone, please re-evaluate this issue.

**Comment by danielwegener** [comment]
@michael-o  I can spontaniously not see why this issue should be pre-java 7 related. Can you help me out? :)

**Comment by plamentotev** [comment]
@danielwegener 
>@michael-o I can spontaniously not see why this issue should be pre-java 7 related. Can you help me out? :)

```java
zipEntry.setTime( lastModified + ( isJava7OrLower ? 1999 : 0 ) );
```

Now Java 7 is the minimum require version so `isJava7OrLower ` was deleted.

**Comment by michael-o** [comment]
@danielwegener For instance stuff like `isJava7OrLower`.
@plamentotev On the second thought, this is less or equal than. This still might be necessary with Java 7+, isn't it?

**Comment by plamentotev** [comment]
@danielwegener @michaelo you're right. My bad, I didn't paid enough
attention, sorry about that. 'isJava7OrLower' is still used of course.

On May 26, 2017 11:24, "Michael Osipov" <notifications@github.com> wrote:

@danielwegener <https://github.com/danielwegener> For instance stuff like
isJava7OrLower.
@plamentotev <https://github.com/plamentotev> On the second thought, this
is less or equal than. This still might be necessary with Java 7+, isn't it?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/codehaus-plexus/plexus-archiver/pull/49#issuecomment-304222963>,
or mute the thread

**Comment by plamentotev** [comment]
I confused it with 'isAtLeastJava7'.

**Comment by hboutemy** [comment]
job done in #118

**Pull request closed without merging**